log4j-api/src/main/java/org/apache/logging/log4j/message/SimpleMessage.java

```diff
@@ -89,17 +90,24 @@ public Object[] getParameters() {
     }
 
     @Override
+    @SuppressWarnings("UndefinedEquals")
     public boolean equals(final Object o) {
         if (this == o) {
             return true;
         }
-        if (o == null || getClass() != o.getClass()) {
+        if (!(o instanceof SimpleMessage)) {
             return false;
         }
 
         final SimpleMessage that = (SimpleMessage) o;
 
-        return !(charSequence != null ? !charSequence.equals(that.charSequence) : that.charSequence != null);
+        /*
+         * https://errorprone.info/bugpattern/UndefinedEquals
+         *
+         * If the char sequences are different, we fall back on string comparison.
+         */
+        return Objects.equals(this.charSequence, that.charSequence)
+                || Objects.equals(this.getFormattedMessage(), that.getFormattedMessage());
```

Comment:
This change makes the `equals` method more useful: two objects are equal if they have the same message.

However it has the side-effect of evaluating `charSequence.toString()`. @jvz, what do you think?

Reply:
How often are messages checked for equality? If it's not a hot path, then I say go for it. `CharSequence::compare` appears to be a Java 11 method, so there's no standard way to compare two char sequences apparently.